### PR TITLE
Separated osx build for develop and releases

### PR DESCRIPTION
**Pull request by serban-nicusor-toptal**
#### Submission Checklist

- [ ] Run unit tests
- Documentation
    - [ ] If a user-facing facing change was made, the documentation PR is here: <LINK>
    - [ ] OR, no user-facing changes were made

## Release notes

After the discussion in https://github.com/stan-dev/math/issues/2623, we've decided to build osx binaries differently for release and develop.

- develop and pull requests will be built on flatiron osx 11.2
- master (nightly) and tags (releases) will be built on gelman osx 10.11.6

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -327,11 +327,13 @@ pipeline {
             }
         }
 
-        stage("Build & test Mac OS X binary") {
+        // Builds on Flatiron osx - 11.2
```
Comment (rok-cesnovar) [inline Jenkinsfile:330]:
```suggestion
        // Builds on Flatiron macOS - recent macOS version
```

File: Jenkinsfile [hunk 2]
```diff
@@ -351,6 +353,31 @@ pipeline {
             post { always { runShell("rm -rf ./*") }}
         }
 
+        // Builds on gelman max osx - 10.11.6
```
Comment (rok-cesnovar) [inline Jenkinsfile:356]:
```suggestion
        // Builds on gelman macOS - version 10.11.6
```

**Review by rok-cesnovar — CHANGES_REQUESTED** [review]
Just some cleanup otherwise looks great. Thanks!

The flatiron machine will update macOS version regularly, so no need to be too specific.

File: Jenkinsfile [hunk 3]
```diff
@@ -327,11 +327,13 @@ pipeline {
...
-        stage("Build & test Mac OS X binary") {
+        // Builds on Flatiron osx - 11.2
+        stage("Build & test Mac OS X binary - develop") {
             when {
                 beforeAgent true
-                expression {
-                    !skipRebuildingBinaries
+                allOf {
+                    expression { !skipRebuildingBinaries }
+                    anyOf { branch 'develop'; changeRequest() }
```
[the remark this answers is not in the document]
Comment (serban-nicusor-toptal) [reply Jenkinsfile:336]: Ah you're right, tho the condition is still valid in our case as we'll be on a PR in the anyOf

**Pull request merged**